### ISSUE TITLE
Problems with DAP support in 4.3.3.1

### ISSUE THREAD

**Issue by durack1** [Unidata/netcdf-c#115]
Hi, I've just hit a problem which appears to be related to DAP tests in 4.3.3.1:

```
> uname -a
Linux machine 2.6.32-504.12.2.el6.x86_64 #1 SMP Sun Feb 1 12:14:02 EST 2015 x86_64 x86_64 x86_64 GNU/Linux
> more /etc/redhat-release
Red Hat Enterprise Linux Workstation release 6.6 (Santiago)
> gcc --version
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-11)
> curl-config --version
libcurl 7.19.7
> curl-config --libs
-lcurl
```
So I've compiled HDF5 1.4.18 and zlib 1.2.8 and am trying to compile netcdf4 using the config args:
```
setenv CPPFLAGS "-I/netcdf-4.3.3.1_install/hdf5+zlib/include"
setenv LDFLAGS "-L/netcdf-4.3.3.1_install/hdf5+zlib/lib"
setenv LIBS -lcurl
./configure --prefix=netcdf-4.3.3.1_install --enable-netcdf-4
```
When I run `make check install` after the configure args above I get the following errors (I've trimmed the full path which includes '.' and '_' characters below for brevity):
```
make[4]: Leaving directory `netcdf-c/ncdap_test'                            
make  check-TESTS                                                                                                                    
make[4]: Entering directory `netcdf-c/ncdap_test'                           
make[5]: Entering directory `netcdf-c/ncdap_test'                           
FAIL: tst_ncdap3.sh                                                                                                                  
FAIL: t_dap3a                                                                                                                        
FAIL: test_cvt3                                                                                                                      
FAIL: test_vara                                                                                                                      
FAIL: tst_remote3.sh                                                                                                                 
FAIL: tst_formatx.sh
FAIL: test_partvar
PASS: testurl.sh
FAIL: test_varm3
PASS: test_nstride_cached
PASS: t_misc
make[6]: Entering directory `netcdf-c/ncdap_test'
echo "#DODSRC" >.dodsrc
make  all-recursive
make[7]: Entering directory `netcdf-c/ncdap_test'
Making all in testdata3
make[8]: Entering directory `netcdf-c/ncdap_test/testdata3'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `netcdf-c/ncdap_test/testdata3'
Making all in expected3
make[8]: Entering directory `netcdf-c/ncdap_test/expected3'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `netcdf-c/ncdap_test/expected3'
Making all in expected4
make[8]: Entering directory `netcdf-c/ncdap_test/expected4'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `netcdf-c/ncdap_test/expected4'
Making all in expectremote3
make[8]: Entering directory `netcdf-c/ncdap_test/expectremote3'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `netcdf-c/ncdap_test/expectremote3'
Making all in expectremote4
make[8]: Entering directory `netcdf-c/ncdap_test/expectremote4'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `netcdf-c/ncdap_test/expectremote4'
make[8]: Entering directory `netcdf-c/ncdap_test'
make[8]: Nothing to be done for `all-am'.
make[8]: Leaving directory `netcdf-c/ncdap_test'
make[7]: Leaving directory `netcdf-c/ncdap_test'
make[6]: Leaving directory `netcdf-c/ncdap_test'
============================================================================
Testsuite summary for netCDF 4.3.3.1
============================================================================
# TOTAL: 11
# PASS:  3
# SKIP:  0
# XFAIL: 0
# FAIL:  8
# XPASS: 0
# ERROR: 0
============================================================================
See ncdap_test/test-suite.log
Please report to support-netcdf@unidata.ucar.edu
============================================================================
make[5]: *** [test-suite.log] Error 1
make[5]: Leaving directory `netcdf-c/ncdap_test'
make[4]: *** [check-TESTS] Error 2
make[4]: Leaving directory `netcdf-c/ncdap_test'
make[3]: *** [check-am] Error 2
make[3]: Leaving directory `netcdf-c/ncdap_test'
make[2]: *** [check-recursive] Error 1
make[2]: Leaving directory `netcdf-c/ncdap_test'
make[1]: *** [check] Error 2
make[1]: Leaving directory `netcdf-c/ncdap_test'
make: *** [check-recursive] Error 1
```

Any ideas what's going wrong here? If I run `./configure --prefix=netcdf-4.3.3.1_install --enable-netcdf-4 --disable-dap` it works fine

**Comment by DennisHeimbigner** [comment]
First, we need to see if you are having problems accessing our
test server. try adding --disable-dap-remote-tests to your ./configure line
and then rebuilding and retesting.  If it succeeds, then you are (or were)
having problems accessing our test server. If it fails, then it is something else.
Also, at the end of the ./configure step, a bunch of settings are listed. Ensure
that those say that dap enables is "yes"

**Comment by durack1** [comment]
Thanks @DennisHeimbigner, here's the output with `--disable-dap-remote-tests`

First from `./configure --prefix=netcdf-4.3.3.1_install --enable-netcdf-4 --disable-dap-remote-tests`

```
configure: netCDF 4.3.3.1                                                                                                            
checking build system type... x86_64-unknown-linux-gnu                                                                               
checking host system type... x86_64-unknown-linux-gnu                                                                                
checking target system type... x86_64-unknown-linux-gnu                                                                              
checking for a BSD-compatible install... /usr/bin/install -c                                                                         
checking whether build environment is sane... yes                                                                                    
checking for a thread-safe mkdir -p... /bin/mkdir -p                                                                                 
checking for gawk... gawk                                                                                                            
checking whether make sets $(MAKE)... yes                                                                                            
checking whether make supports nested variables... yes                                                                               
checking whether to enable maintainer-specific portions of Makefiles... no                                                           
configure: checking user options                                                                                                     
checking whether a win32 DLL is desired... no                                                                                        
checking whether a NCIO_MINBLOCKSIZE was specified... 256                                                                            
checking if fsync support is enabled... no                                                                                           
checking if jna bug workaround is enabledd... no                                                                                     
checking whether extra valgrind tests should be run... no                                                                            
checking whether we should try to build netCDF-4... yes                                                                              
checking do we require hdf5 dynamic-loading support... yes                                                                           
checking whether reading of HDF4 SD files is to be enabled... no                                                                     
checking whether to fetch some sample HDF4 files from Unidata ftp site to test HDF4 reading (requires wget)... no                    
checking whether we should attempt to install netcdf-fortran (EXPERIMENTAL)... no                                                    
checking whether parallel I/O for classic and 64-bit offset files using parallel-netcdf is to be enabled... no                       
checking whether extra example tests should be run... no                                                                             
checking whether parallel IO tests should be run... no                                                                               
checking whether a default chunk size in bytes was specified... 4194304                                                              
checking whether a maximum per-variable cache size for HDF5 was specified... 67108864                                                
checking whether a number of chunks for the default per-variable cache was specified... 10                                           
checking whether a default file cache size for HDF5 was specified... 4194304                                                         
checking whether a default file cache maximum number of elements for HDF5 was specified... 1009                                      
checking whether a default cache preemption for HDF5 was specified... 0.75                                                           
checking whether netCDF-4 logging is enabled... no                                                                                   
checking whether cdmremote client is to be built... no                                                                               
checking whether DAP client is to be built... yes                                                                                    
checking for style of include used by make... GNU                                                                                    
checking for gcc... gcc                                                                                                              
checking whether the C compiler works... yes                                                                                         
checking for C compiler default output file name... a.out                                                                            
checking for suffix of executables...                                                                                                
checking whether we are cross compiling... no                                                                                        
checking for suffix of object files... o                                                                                             
checking whether we are using the GNU C compiler... yes                                                                              
checking whether gcc accepts -g... yes                                                                                               
checking for gcc option to accept ISO C89... none needed                                                                             
checking whether gcc understands -c and -o together... yes                                                                           
checking dependency style of gcc... gcc3                                                                                             
checking for curl_easy_setopt in -lcurl... yes                                                                                       
checking for library containing curl_easy_setopt... none required                                                                    
checking whether dap remote testing should be enabled (default on)... no                                                             
checking whether dap remote authorization testing should be enabled (default off)... no                                              
checking whether netcdf4 group names should be enabled (default on)... yes                                                           
checking whether the time-consuming dap tests should be enabled (default off)... no                                                  
checking whether netCDF extra tests should be run (developers only)... no                                                            
checking whether FFIO will be used... no                                                                                             
checking whether examples should be built... yes                                                                                     
checking whether v2 netCDF API should be built... yes                                                                                
checking whether the ncgen/ncdump/nccopy should be built... yes                                                                      
checking whether test should be built and run... yes                                                                                 
checking whether large file (> 2GB) tests should be run... no                                                                        
checking whether benchmaks should be run (experimental)... no                                                                        
checking whether extreme numbers should be used in tests... yes                                                                      
checking where to put large temp files if large file tests are run... .                                                              
configure: finding C compiler                                                                                                        
checking for gcc... (cached) gcc                                                                                                     
checking whether we are using the GNU C compiler... (cached) yes                                                                     
checking whether gcc accepts -g... (cached) yes                                                                                      
checking for gcc option to accept ISO C89... (cached) none needed                                                                    
checking whether gcc understands -c and -o together... (cached) yes                                                                  
checking dependency style of gcc... (cached) gcc3                                                                                    
checking for an ANSI C-conforming const... yes                                                                                       
checking whether CURLOPT_USERNAME is defined... yes                                                                                  
checking whether CURLOPT_PASSWORD is defined... yes                                                                                  
checking whether CURLOPT_KEYPASSWD is defined... yes                                                                                 
checking whether CURLINFO_RESPONSE_CODE is defined... yes                                                                            
checking whether CURLOPT_CHUNK_BGN_FUNCTION is defined... no                                                                         
configure: setting up libtool                                                                                                        
checking how to print strings... printf                                                                                              
checking for a sed that does not truncate output... /bin/sed                                                                         
checking for grep that handles long lines and -e... /bin/grep                                                                        
checking for egrep... /bin/grep -E                                                                                                   
checking for fgrep... /bin/grep -F                                                                                                   
checking for ld used by gcc... /usr/bin/ld                                                                                           
checking if the linker (/usr/bin/ld) is GNU ld... yes                                                                                
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B                                                                
checking the name lister (/usr/bin/nm -B) interface... BSD nm                                                                        
checking whether ln -s works... yes                                                                                                  
checking the maximum length of command line arguments... 1966080                                                                     
checking whether the shell understands some XSI constructs... yes                                                                    
checking whether the shell understands "+="... yes                                                                                   
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop             
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop                            
checking for /usr/bin/ld option to reload object files... -r                                                                         
checking for objdump... objdump                                                                                                      
checking how to recognize dependent libraries... pass_all                                                                            
checking for dlltool... no                                                                                                           
checking how to associate runtime and link libraries... printf %s\n                                                                  
checking for ar... ar                                                                                                                
checking for archiver @FILE support... @                                                                                             
checking for strip... strip                                                                                                          
checking for ranlib... ranlib                                                                                                        
checking command to parse /usr/bin/nm -B output from gcc object... ok                                                                
checking for sysroot... no                                                                                                           
checking for mt... no                                                                                                                
checking if : is a manifest tool... no                                                                                               
checking how to run the C preprocessor... gcc -E                                                                                     
checking for ANSI C header files... yes                                                                                              
checking for sys/types.h... yes                                                                                                      
checking for sys/stat.h... yes                                                                                                       
checking for stdlib.h... yes                                                                                                         
checking for string.h... yes                                                                                                         
checking for memory.h... yes                                                                                                         
checking for strings.h... yes                                                                                                        
checking for inttypes.h... yes                                                                                                       
checking for stdint.h... yes                                                                                                         
checking for unistd.h... yes                                                                                                         
checking for dlfcn.h... yes                                                                                                          
checking for objdir... .libs                                                                                                         
checking if gcc supports -fno-rtti -fno-exceptions... no                                                                             
checking for gcc option to produce PIC... -fPIC -DPIC                                                                                
checking if gcc PIC flag -fPIC -DPIC works... yes                                                                                    
checking if gcc static flag -static works... no                                                                                      
checking if gcc supports -c -o file.o... yes                                                                                         
checking if gcc supports -c -o file.o... (cached) yes                                                                                
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes                                         
checking whether -lc should be explicitly linked in... no                                                                            
checking dynamic linker characteristics... GNU/Linux ld.so                                                                           
checking how to hardcode library paths into programs... immediate                                                                    
checking whether stripping libraries is possible... yes                                                                              
checking if libtool supports shared libraries... yes                                                                                 
checking whether to build shared libraries... yes                                                                                    
checking whether to build static libraries... yes                                                                                    
configure: finding other utilities                                                                                                   
checking for m4... m4                                                                                                                
checking for doxygen... doxygen                                                                                                      
checking for dot... no                                                                                                               
configure: WARNING: dot not found - will use simple charts in documentation                                                          
checking for special C compiler options needed for large files... no                                                                 
checking for _FILE_OFFSET_BITS value needed for large files... no                                                                    
configure: displaying some results                                                                                                   
checking CPPFLAGS... -I/export/durack1/bin/downloads/netcdf4.3.3.1/netcdf-4.3.3.1_install/hdf5+zlib/include                          
checking CC CFLAGS... gcc -g -O2                                                                                                     
checking type gcc... gcc is /usr/bin/gcc                                                                                             
checking CXX... unset                                                                                                                
checking FC... unset                                                                                                                 
checking F90... unset                                                                                                                
checking AR... ar                                                                                                                    
checking AR_FLAGS... cru                                                                                                             
checking type ar... ar is /usr/bin/ar                                                                                                
checking NM... /usr/bin/nm -B                                                                                                        
checking NMFLAGS...                                                                                                                  
configure: checking types, headers, and functions                                                                                    
checking for stdlib.h... (cached) yes                                                                                                
checking for sys/types.h... (cached) yes                                                                                             
checking sys/param.h usability... yes                                                                                                
checking sys/param.h presence... yes                                                                                                 
checking for sys/param.h... yes                                                                                                      
checking for dirent.h that defines DIR... yes                                                                                        
checking for library containing opendir... none required                                                                             
checking for ANSI C header files... (cached) yes                                                                                     
checking for sys/wait.h that is POSIX.1 compatible... yes                                                                            
checking whether time.h and sys/time.h may both be included... yes                                                                   
checking locale.h usability... yes                                                                                                   
checking locale.h presence... yes                                                                                                    
checking for locale.h... yes                                                                                                         
checking stdio.h usability... yes                                                                                                    
checking stdio.h presence... yes                                                                                                     
checking for stdio.h... yes                                                                                                          
checking stdarg.h usability... yes                                                                                                   
checking stdarg.h presence... yes                                                                                                    
checking for stdarg.h... yes                                                                                                         
checking errno.h usability... yes                                                                                                    
checking errno.h presence... yes                                                                                                     
checking for errno.h... yes                                                                                                          
checking ctype.h usability... yes                                                                                                    
checking ctype.h presence... yes                                                                                                     
checking for ctype.h... yes                                                                                                          
checking fcntl.h usability... yes                                                                                                    
checking fcntl.h presence... yes                                                                                                     
checking for fcntl.h... yes                                                                                                          
checking malloc.h usability... yes                                                                                                   
checking malloc.h presence... yes                                                                                                    
checking for malloc.h... yes                                                                                                         
checking for stdlib.h... (cached) yes                                                                                                
checking for string.h... (cached) yes                                                                                                
checking for strings.h... (cached) yes                                                                                               
checking for unistd.h... (cached) yes                                                                                                
checking for sys/stat.h... (cached) yes                                                                                              
checking getopt.h usability... yes                                                                                                   
checking getopt.h presence... yes                                                                                                    
checking for getopt.h... yes                                                                                                         
checking sys/time.h usability... yes                                                                                                 
checking sys/time.h presence... yes                                                                                                  
checking for sys/time.h... yes                                                                                                       
checking for vprintf... yes                                                                                                          
checking for _doprnt... no                                                                                                           
checking sys/resource.h usability... yes                                                                                             
checking sys/resource.h presence... yes                                                                                              
checking for sys/resource.h... yes                                                                                                   
checking for stdbool.h that conforms to C99... yes                                                                                   
checking for _Bool... yes                                                                                                            
checking for strlcat... no                                                                                                           
checking for strerror... yes                                                                                                         
checking for snprintf... yes                                                                                                         
checking for strchr... yes                                                                                                           
checking for strrchr... yes                                                                                                          
checking for strcat... yes                                                                                                           
checking for strcpy... yes                                                                                                           
checking for strdup... yes                                                                                                           
checking for strcasecmp... yes                                                                                                       
checking for strtod... yes                                                                                                           
checking for strtoll... yes                                                                                                          
checking for strtoull... yes                                                                                                         
checking for strstr... yes                                                                                                           
checking for mkstemp... yes                                                                                                          
checking for rand... yes                                                                                                             
checking for memcmp... yes                                                                                                           
checking for getrlimit... yes                                                                                                        
checking for gettimeofday... yes                                                                                                     
checking for fsync... yes                                                                                                            
checking for MPI_Comm_f2c... no                                                                                                      
checking whether in-memory files are enabled... configure: --enable-dap requires --enable-diskless                                   
yes                                                                                                                                  
checking for memmove... yes                                                                                                          
checking for getpagesize... yes                                                                                                      
checking for sysconf... yes                                                                                                          
checking whether mmap is enabled for in-memory files... no                                                                           
checking for stdlib.h... (cached) yes                                                                                                
checking for unistd.h... (cached) yes                                                                                                
checking for sys/param.h... (cached) yes                                                                                             
checking for getpagesize... (cached) yes                                                                                             
checking for working mmap... yes                                                                                                     
checking for mremap... yes                                                                                                           
checking for size_t... yes                                                                                                           
checking for working alloca.h... yes                                                                                                 
checking for alloca... yes                                                                                                           
checking whether isnan is declared... yes                                                                                            
checking whether isinf is declared... yes                                                                                            
checking whether isfinite is declared... yes                                                                                         
checking whether signbit is declared... yes                                                                                          
checking for struct stat.st_blksize... yes                                                                                           
checking for IEEE floating point format... yes                                                                                       
checking for size_t... (cached) yes                                                                                                  
checking for off_t... yes                                                                                                            
checking for ssize_t... yes                                                                                                          
checking for ptrdiff_t... yes                                                                                                        
checking for uchar... no                                                                                                             
checking for longlong... no                                                                                                          
checking whether char is unsigned... no                                                                                              
checking whether byte ordering is bigendian... no                                                                                    
checking size of short... 2                                                                                                          
checking size of int... 4                                                                                                            
checking size of long... 8                                                                                                           
checking size of long long... 8                                                                                                      
checking size of float... 4                                                                                                          
checking size of double... 8                                                                                                         
checking size of off_t... 8                                                                                                          
checking size of size_t... 8                                                                                                         
checking for library containing deflate... none required                                                                             
checking for floor in -lm... yes                                                                                                     
checking for dlfcn.h... (cached) yes                                                                                                 
checking for dlopen in -ldl... yes                                                                                                   
checking for library containing H5Fflush... -lhdf5                                                                                   
checking for library containing H5DSis_scale... -lhdf5_hl                                                                            
checking hdf5.h usability... yes                                                                                                     
checking hdf5.h presence... yes                                                                                                      
checking for hdf5.h... yes                                                                                                           
checking for H5Pget_fapl_mpiposix... no                                                                                              
checking for H5Pget_fapl_mpio... no                                                                                                  
checking for H5Pset_deflate... yes                                                                                                   
checking for H5Z_SZIP... no                                                                                                          
checking whether parallel I/O features are to be included... no                                                                      
checking for unsigned long long int... yes                                                                                           
checking for long long int... no                                                                                                     
checking for unsigned long long int... (cached) yes                                                                                  
checking what to call the output of the ftpbin target... binary-netcdf-4.3.3.1.tar                                                   
checking value of LIBS... -lhdf5_hl -lhdf5 -ldl -lm -lcurl                                                                           
checking If compilation is for use with JNA... no                                                                                    
configure: generating header files and makefiles                                                                                     
checking that generated files are newer than configure... done                                                                       
configure: creating ./config.status                                                                                                  
config.status: creating Makefile                                                                                                     
config.status: creating nc-config                                                                                                    
config.status: creating netcdf.pc                                                                                                    
config.status: creating libnetcdf.settings                                                                                           
config.status: creating postinstall.sh                                                                                               
config.status: creating include/netcdf_meta.h                                                                                        
config.status: creating include/Makefile                                                                                             
config.status: creating h5_test/Makefile
config.status: creating docs/Makefile
config.status: creating docs/images/Makefile
config.status: creating libsrc/Makefile
config.status: creating libsrc4/Makefile
config.status: creating libsrc5/Makefile
config.status: creating nctest/Makefile
config.status: creating nc_test4/Makefile
config.status: creating nc_test/Makefile
config.status: creating ncdump/Makefile
config.status: creating ncgen3/Makefile
config.status: creating ncgen/Makefile
config.status: creating examples/Makefile
config.status: creating examples/C/Makefile
config.status: creating examples/CDL/Makefile
config.status: creating oc2/Makefile
config.status: creating libdap2/Makefile
config.status: creating libdispatch/Makefile
config.status: creating liblib/Makefile
config.status: creating ncdump/cdl/Makefile
config.status: creating ncdump/expected/Makefile
config.status: creating ncdap_test/Makefile
config.status: creating ncdap_test/testdata3/Makefile
config.status: creating ncdap_test/expected3/Makefile
config.status: creating ncdap_test/expected4/Makefile
config.status: creating ncdap_test/expectremote3/Makefile
config.status: creating ncdap_test/expectremote4/Makefile
config.status: creating config.h
config.status: executing depfiles commands
config.status: executing libtool commands
# NetCDF C Configuration Summary
==============================

# General
-------
NetCDF Version:         4.3.3.1
Configured On:          Wed Apr 22 11:36:47 PDT 2015
Host System:            x86_64-unknown-linux-gnu
Build Directory:        /export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c
Install Prefix:         /export/durack1/bin/downloads/netcdf4.3.3.1/netcdf-4.3.3.1_install

# Compiling Options
-----------------
C Compiler:             /usr/bin/gcc
CFLAGS:                 -g -O2
CPPFLAGS:               -I/export/durack1/bin/downloads/netcdf4.3.3.1/netcdf-4.3.3.1_install/hdf5+zlib/include
LDFLAGS:                -L/export/durack1/bin/downloads/netcdf4.3.3.1/netcdf-4.3.3.1_install/hdf5+zlib/lib
AM_CFLAGS:
AM_CPPFLAGS:
AM_LDFLAGS:
Shared Library:         yes
Static Library:         yes
Extra libraries:        -lhdf5_hl -lhdf5 -ldl -lm -lcurl

# Features
--------
NetCDF-2 API:           yes
NetCDF-4 API:           yes
HDF4 Support:           no
HDF5 Support:           yes
PNetCDF Support:        no
Parallel Support:       no
DAP Support:            yes
Diskless Support:       yes
MMap Support:           no
JNA Support:            no
```
And the errors when running `make check install`:
```
make[4]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'                            
make  check-TESTS                                                                                                                    
make[4]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'                           
make[5]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'                           
FAIL: tst_ncdap3.sh
FAIL: t_dap3a
FAIL: test_cvt3
FAIL: test_vara
make[6]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
echo "#DODSRC" >.dodsrc
make  all-recursive
make[7]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
Making all in testdata3
make[8]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/testdata3'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/testdata3'
Making all in expected3
make[8]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expected3'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expected3'
Making all in expected4
make[8]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expected4'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expected4'
Making all in expectremote3
make[8]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expectremote3'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expectremote3'
Making all in expectremote4
make[8]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expectremote4'
make[8]: Nothing to be done for `all'.
make[8]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/expectremote4'
make[8]: Entering directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[8]: Nothing to be done for `all-am'.
make[8]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[7]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[6]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
============================================================================
Testsuite summary for netCDF 4.3.3.1
============================================================================
# TOTAL: 4
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  4
# XPASS: 0
# ERROR: 0
============================================================================
See ncdap_test/test-suite.log
Please report to support-netcdf@unidata.ucar.edu
============================================================================
make[5]: *** [test-suite.log] Error 1
make[5]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[4]: *** [check-TESTS] Error 2
make[4]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[3]: *** [check-am] Error 2
make[3]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[2]: *** [check-recursive] Error 1
make[2]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make[1]: *** [check] Error 2
make[1]: Leaving directory `/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test'
make: *** [check-recursive] Error 1
```

**Comment by DennisHeimbigner** [comment]
Ok, so the problem is something else.
If possible, go into the directory nc_dap2
and execute t_dap3a, test_cvt3, and test_vara
directly on the command line e.,g. cd nc_dap2; ./t_dap3a
and see if it produces any useful output.

**Comment by durack1** [comment]
@DennisHeimbigner nope, doesn't look very useful:
```
[durack1@oceanonly netcdf-c]$ cd ncdap_test/
netcdf-c/ncdap_test
[durack1@oceanonly ncdap_test]$ ls -l
total 676
-rw------- 1 durack1 climate   2398 Apr 22 11:48 CMakeLists.txt
drwx------ 2 durack1 climate   4096 Apr 22 11:58 expected3
drwx------ 2 durack1 climate   4096 Apr 22 11:58 expected4
drwx------ 2 durack1 climate   4096 Apr 22 11:58 expectremote3
drwx------ 2 durack1 climate   4096 Apr 22 11:58 expectremote4
drwx------ 2 durack1 climate   4096 Apr 22 11:48 expecttds3
-rw------- 1 durack1 climate  47140 Apr 22 11:58 Makefile
-rw------- 1 durack1 climate   2783 Apr 22 11:48 Makefile.am
-rw------- 1 durack1 climate  48708 Apr 22 11:48 Makefile.in
-rw------- 1 durack1 climate    883 Apr 22 11:48 nctestserver.c
drwx------ 2 durack1 climate   4096 Apr 22 11:48 nocacheremote3
drwx------ 2 durack1 climate   4096 Apr 22 11:48 nocacheremote4
drwx------ 2 durack1 climate   4096 Apr 22 12:03 results
-rwx------ 1 durack1 climate   1714 Apr 22 11:48 run_ncdap_tests.sh
drwx------ 2 durack1 climate   4096 Apr 22 11:48 special3
-rw------- 1 durack1 climate   4599 Apr 22 11:48 t_auth.c
-rwx------ 1 durack1 climate   9010 Apr 22 12:03 t_dap3a
-rw------- 1 durack1 climate  14927 Apr 22 12:03 t_dap3a.c
-rw------- 1 durack1 climate    189 Apr 22 12:03 t_dap3a.log
-rw------- 1 durack1 climate  27176 Apr 22 12:03 t_dap3a.o
-rw------- 1 durack1 climate     84 Apr 22 12:03 t_dap3a.trs
-rw------- 1 durack1 climate   1226 Apr 22 11:48 t_dap3b.c
-rw------- 1 durack1 climate   1172 Apr 22 11:48 t_dap3c.c
-rw------- 1 durack1 climate  29775 Apr 22 11:48 t_dap4a.c
-rw------- 1 durack1 climate  14868 Apr 22 11:48 t_dap4.c
-rw------- 1 durack1 climate   1173 Apr 22 11:48 t_dap4c.c
-rw------- 1 durack1 climate  14907 Apr 22 11:48 t_dap.c
-rwx------ 1 durack1 climate   4011 Apr 22 11:48 testauth.sh
-rwx------ 1 durack1 climate   9031 Apr 22 12:03 test_cvt3
-rw------- 1 durack1 climate    190 Apr 22 12:03 test_cvt3.log
-rw------- 1 durack1 climate     84 Apr 22 12:03 test_cvt3.trs
-rw------- 1 durack1 climate  15089 Apr 22 11:48 test_cvt.c
-rw------- 1 durack1 climate  27240 Apr 22 12:03 test_cvt.o
drwx------ 2 durack1 climate  12288 Apr 22 11:58 testdata3
-rw------- 1 durack1 climate   7455 Apr 22 11:48 test_nstride_cached.c
-rw------- 1 durack1 climate 104177 Apr 22 11:48 test_partvar2.c
-rw------- 1 durack1 climate  24476 Apr 22 11:48 test_partvar.c
-rw------- 1 durack1 climate   1725 Apr 22 12:03 test-suite.log
-rwx------ 1 durack1 climate   2066 Apr 22 11:48 testurl.sh
-rwx------ 1 durack1 climate   9032 Apr 22 12:03 test_vara
-rw------- 1 durack1 climate  25895 Apr 22 11:48 test_vara.c
-rw------- 1 durack1 climate    144 Apr 22 12:03 test_vara.log
-rw------- 1 durack1 climate  16472 Apr 22 12:03 test_vara.o
-rw------- 1 durack1 climate     84 Apr 22 12:03 test_vara.trs
-rw------- 1 durack1 climate   6615 Apr 22 11:48 test_varm3.c
-rw------- 1 durack1 climate    903 Apr 22 11:48 t_misc.c
-rwx------ 1 durack1 climate    583 Apr 22 11:48 tst_formatx.sh
-rwx------ 1 durack1 climate    558 Apr 22 11:48 tst_longremote3.sh
-rwx------ 1 durack1 climate    546 Apr 22 11:48 tst_longremote4.sh
-rwx------ 1 durack1 climate    831 Apr 22 11:48 tst_ncdap3.sh
-rw------- 1 durack1 climate    817 Apr 22 12:03 tst_ncdap3.sh.log
-rw------- 1 durack1 climate     84 Apr 22 12:03 tst_ncdap3.sh.trs
-rwx------ 1 durack1 climate    617 Apr 22 11:48 tst_ncdap4.sh
-rwx------ 1 durack1 climate   2556 Apr 22 11:48 tst_ncdap.sh
-rwx------ 1 durack1 climate   1762 Apr 22 11:48 tst_ncdap_shared.sh
-rwx------ 1 durack1 climate    590 Apr 22 11:48 tst_nocache3.sh
-rwx------ 1 durack1 climate    590 Apr 22 11:48 tst_nocache4.sh
-rwx------ 1 durack1 climate    543 Apr 22 11:48 tst_remote3.sh
-rwx------ 1 durack1 climate    583 Apr 22 11:48 tst_remote4.sh
-rwx------ 1 durack1 climate   8866 Apr 22 11:48 tst_remote.sh
-rwx------ 1 durack1 climate   5509 Apr 22 11:48 tst_special.sh
-rwx------ 1 durack1 climate   3576 Apr 22 11:48 tst_tds.sh
[durack1@oceanonly ncdap_test]$ t_dap3a
*** FAIL: $abs_top_srcdir not defined: location= t_dap3a.c:127
[durack1@oceanonly ncdap_test]$ test_cvt3
*** FAIL: $abs_top_srcdir not defined: location= test_cvt.c:126
[durack1@oceanonly ncdap_test]$ test_vara
*** FAIL: $abs_top_srcdir not defined: location= test_vara.c:84
```

**Comment by DennisHeimbigner** [comment]
ok, lets focus on t_vara
First, we need to set the env variable TOPSRCDIR.
So assuming you are using bash or sh as your shell,
do this cmd:
    export TOPSRCDIR="<path>"
where <path> is the absolute path for the netcdf-c directory.
e.g. /x/y/netcdf-c
Then rerun t_vara and see if anything useful comes out.

**Comment by durack1** [comment]
@DennisHeimbigner we're getting somewhere..

```
[netcdf-c]$ setenv TOPSRCDIR /export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c
[netcdf-c]$ cd ncdap_test/
/export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test
[ncdap_test]$ test_vara
test_vara: url=file:///export/durack1/bin/downloads/netcdf4.3.3.1/150422_build/netcdf-c/ncdap_test/testdata3/test.06
*** PASS
```

**Comment by DennisHeimbigner** [comment]
Ok, that is surprising; It almost looks like an autoconf problem.
Are you, by any chance, in a position to try building with Cmake?

**Comment by durack1** [comment]
@DennisHeimbigner I think that's a possibility.. Will v2.8.9 be ok?
```
[ncdap_test]$ which cmake
/usr/local/bin/cmake
[ncdap_test]$ cmake --version
cmake version 2.8.9
```

**Comment by DennisHeimbigner** [comment]
According to CMakeLists.txt, version 2.8.12 or later is required.

**Comment by durack1** [comment]
@DennisHeimbigner ok so we're good to go..

```
[downloads]$ which cmake
/export/durack1/bin/downloads/cmake3.2.2/150422_build/bin/cmake
[downloads]$ cmake --version
cmake version 3.2.2
```

**Comment by durack1** [comment]
@DennisHeimbigner is there guidance available to build netcdf-c with cmake?

**Comment by DennisHeimbigner** [comment]
See if the attached is clear enough.
=Dennis Heimbigner
  Unidata


Paul J. Durack wrote:
> @DennisHeimbigner is there guidance available to build netcdf-c with cmake?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Unidata/netcdf-c/issues/115#issuecomment-95728924

Getting and Building netCDF {#getting_and_building_netcdf}
=============================

[TOC]

This document is for getting and building the netCDF C library and
utilities, version 4.3.0.  Other libraries that depend on the netCDF C
library, such as the Fortran and C++ libraries, are available as
separate distributions that can be built and installed after the C
library is successfully installed.  The netCDF-Java library is also a
separate distribution that is currently independent of the netCDF C
library.


Getting netCDF-C {#getting}
=========================

* For information regarding the netCDF-Fortran libraries, see \subpage building_netcdf_fortran.

Getting pre-built netCDF-C libraries. {#sec_get_pre_built}
-------------------------------------

The easiest way to get netCDF is through a package management program,
such as rpm, yum, adept, and others. NetCDF is available from many
different repositories, including the default Red Hat and Ubuntu
repositories.

When getting netCDF from a software repository, you will wish to get
the development version of the package ("netcdf-devel"). This includes
the netcdf.h header file.

Pre-release libraries for Windows may be found here: \ref winbin.

Getting the latest netCDF-C Source Code {#sec_get_source}
----------------------------------------

Starting with netCDF-C version 4.3.1, the netCDF-C source code is hosted at the
Unidata GitHub repository, available at http://github.com/Unidata/netcdf-c.

Two options are available for building from source:

- The latest release.
- The developer snapshot.

### The latest release {#sec_latest_release}

The latest release may be downloaded from github at the following location:

- http://github.com/Unidata/netcdf-c/releases

Source files are available in `.tar.gz` and `.zip` formats.

### The developer snapshot {#sec_dev_snapshot}

The developer snapshot may be cloned from github directly by using the `git` command.

> $ git clone http://github.com/Unidata/netcdf-c netcdf-c

**Note:**

> ***The developer snapshot release contains bug-fixes and new
features added since the last full release. It may also contain
portability bugs.***

Once you have downloaded and unpacked the distribution, see the
following section on \ref building.

Building netCDF-C {#building}
===========================

The netCDF-C library and utilities require third-party libraries for
full functionality. (See \ref architecture).
- \ref build_default
- \ref build_classic
- \ref build_hdf4
- \ref build_parallel
- \ref building_netcdf_fortran
- \ref configure_options

Requirements {#netcdf_requirements}
----------------------------------

* HDF5 1.8.9 (netcdf-4 support)
* zlib 1.2.5
* curl 7.18.0 (DAP support)


CMake and Windows support {#sub}
--------------------------------

- \ref netCDF-CMake
- \subpage winbin

Building with netCDF-4 and the Remote Data Client {#build_default}
--------------------------------

The usual way of building netCDF requires the HDF5, zlib, and curl
libraries. (And, optionally, the szlib library). Versions required are
at least HDF5 1.8.9, zlib 1.2.5, and curl 7.18.0 or later.
(Optionally, if building with szlib, get szip 2.0 or later.)

HDF5 1.8.9 and zlib 1.2.7 packages are available from the <a
href="ftp://ftp.unidata.ucar.edu/pub/netcdf/netcdf-4">netCDF-4 ftp
site</a>. If you wish to use the remote data client code, then you
will also need libcurl, which can be obtained from the <a
href="http://curl.haxx.se/download.html">curl website</a>.

Make sure you run ``make check'' for the HDF5 and zlib
distributions. They are very well-behaved distributions, but sometimes
the build doesn't work (perhaps because of something subtly
misconfigured on the target machine). If one of these libraries is not
working, netCDF will have serious problems.

Note that for building netCDF, it is not necessary to build the HDF5
Fortran, C++, or Java API's.  Only the HDF5 C library is used.

Optionally, you can also build netCDF-4 with the szip library
(a.k.a. szlib). NetCDF cannot create szipped data files, but can read
HDF5 data files that have used szip.
8
There are license restrictions on the use of szip, see the section on
licensing terms in the <a
href="http://www.hdfgroup.org/doc_resource/SZIP/">web page on szip
compression in HDF products</a>. These license restrictions seem to
apply to commercial users who are writing data. (Data readers are not
restricted.) But here at NetCDF World Headquarters, in Sunny Boulder,
Colorado, there are no lawyers, only programmers, so please read the
szip documents for the license agreement to see how it applies to your
situation.

If ``make check'' fails for either zlib or HDF5, the problem must be
resolved before the netCDF-4 installation can continue. For HDF5
problems, see the <a
href="http://www.hdfgroup.org/services/support.html">HDF5 help
services</a>.

Build zlib like this:

~~~
$ ./configure --prefix=/home/username/local
$ make check install
~~~

Then you build HDF5, specifying the location of the zlib library:

~~~
$ ./configure --with-zlib=/home/username/local --prefix=/home/username/local
$ make check install
~~~

In all cases, the installation location specified with the <CODE>--prefix</CODE>
option must be different from the source directory where the software
is being built.

Note that for shared libraries, you may need to add the install
directory to the LD_LIBRARY_PATH environment variable. See
the <a href="http://www.unidata.ucar.edu/netcdf/docs/faq.html#Shared%20Libraries">netCDF
FAQ</a> for more details on using shared libraries.

If you are building HDF5 with szip, then include the <CODE>--with-szlib=</CODE>
option, with the directory holding the szip library.

After HDF5 is done, build netcdf, specifying the location of the
HDF5, zlib, and (if built into HDF5) the szip header files and
libraries in the CPPFLAGS and LDFLAGS environment variables. For example:

~~~
$ CPPFLAGS=-I/home/username/local/include LDFLAGS=-L/home/username/local/lib ./configure --prefix=/home/username/local
$ make check install
~~~

The configure script will try to find necessary tools in your
path. When you run configure you may optionally use the <CODE>--prefix</CODE>
argument to change the default installation directory. The above
examples install the zlib, HDF5, and netCDF-4 libraries in
/home/username/local/lib, the header file in /home/username/local/include, and the
utilities in /home/username/local/bin. If you don't provide a <CODE>--prefix</CODE>
option, installation will be in /usr/local/, in subdirectories lib/,
include/, and bin/.  The installation location specified with the
<CODE>--prefix</CODE> option must be different from the source directory where the
software is being built.

Building netCDF with Classic Library Only {#build_classic}
---------------------------------------

It is possible to build the netCDF C libraries and utilities so that
only the netCDF classic and 64-bit offset formats are supported, or
the remote data access client is not built. (See \ref netcdf_format)
for more information about the netCDF format variants.  See the <a
href="http://opendap.org/netCDF-DAP">netCDF-DAP site</a>
for more information about remote client access to data
on OPeNDAP servers.)

To build without support for the netCDF-4 formats or the additional
netCDF-4 functions, but with remote access, use:

~~~
$ ./configure --prefix=/home/username/local --disable-netcdf-4
$ make check install
~~~

(Replace `/home/username/local` with the name of the directory where
netCDF is to be installed.  The installation location specified with
the <CODE>--prefix</CODE> option must be different from the source directory where
the software is being built.)

Starting with version 4.1.1 the netCDF C libraries and utilities have
supported remote data access, using the OPeNDAP protocols.  To build
with full support for netCDF-4 APIs and format but without remote
client access, use:

~~~
$ ./configure --prefix=/home/username/local --disable-dap
$ make check install
~~~

To build without netCDF-4 support or remote client access, use:

~~~
$ ./configure --prefix=/home/username/local --disable-netcdf-4 --disable-dap
$ make check install
~~~

If you get the message that netCDF installed correctly, then you are
done!

Building with HDF4 Support {#build_hdf4}
---------------------

The netCDF-4 library can (since version 4.1) read HDF4 data files, if
they were created with the SD (Scientific Data) API.

For this to work, you must build the HDF4 library with the
configure option
~~~
  --disable-netcdf
~~~
to prevent it from building an HDF4 version of the netCDF-2 library
that conflicts with the netCDF-2 functions that are built into the Unidata
netCDF library.

Then, when building netCDF-4, use the
~~~
  --enable-hdf4
~~~
option to configure. The location for the HDF4 header files and
library must be set in the CPPFLAGS and LDFLAGS options.

For HDF4 access to work, the library must be build with netCDF-4
features.

Here's an example, assuming the HDF5 library has been built and
installed in H5DIR and you will build and install the HDF4 library in
H4DIR (which could be the same as H5DIR):

~~~
# Build and install HDF4
$ cd ${HDF4_SOURCE_DIRECTORY}
$ ./configure --enable-shared --disable-netcdf --disable-fortran --prefix=${H4DIR}
$ make
$ make install
$ # Build and install netCDF with HDF4 access enabled
$ cd ${NETCDF_SOURCE_DIRECTORY}
$ CPPFLAGS="-I${H5DIR}/include -I${H4DIR}/include" \
$          LDFLAGS="-L${H5DIR}/lib -L${H4DIR}/lib" \
$ ./configure --enable-hdf4 --enable-hdf4-file-tests
$ make check
$ make install
~~~

Building with Parallel I/O Support {#build_parallel}
--------------

For parallel I/O to work, HDF5 must be installed with
–enable-parallel, and an MPI library (and related libraries) must be
made available to the HDF5 configure. This can be accomplished with
the mpicc wrapper script, in the case of MPICH2.

The following works to build HDF5 with parallel I/O on our netCDF
testing system:

~~~
CC=mpicc ./configure --enable-parallel
make check install
~~~

If the HDF5 used by netCDF has been built with parallel I/O, then
netCDF will also be built with support for parallel I/O. This allows
parallel I/O access to netCDF-4/HDF5 files. Note that shared libraries
are not supported for parallel HDF5, which makes linking more
difficult to get right.  "LIBS=-ldl" is also sometimes needed to link
successfully with parallel HDF5 libraries.
(See /ref netcdf_formats for more information about the netCDF format
variants.)

The following works to build netCDF-4 with parallel I/O on our netCDF
testing system:

~~~
$ H5DIR=/where/parallel/HDF5/was/installed
$ CPPFLAGS="-I${H5DIR}/include"
$ CC=mpicc
$ LDFLAGS=-L${H5DIR}/lib
$ LIBS=-ldl
$ ./configure --disable-shared --enable-parallel-tests
$ make check install
~~~


If parallel I/O access to netCDF classic and 64-bit offset files is
also needed, the parallel-netcdf library should also be installed.
(Note: the previously recommended <a
href=ftp://ftp.unidata.ucar.edu/pub/netcdf/contrib/pnetcdf.h>replacement
pnetcdf.h</a> should no longer be used.)  Then configure netCDF with the
"--enable-pnetcdf" option.

Linking to netCDF-C {#linking}
-------------------

For static build, to use netCDF-4 you must link to all the libraries,
netCDF, HDF5, zlib, szip (if used with HDF5 build), and curl (if the
remote access client has not been disabled). This will mean -L options
to your build for the locations of the libraries, and -l (lower-case
L) for the names of the libraries.

For example, one user reports that she can build other applications
with netCDF-4 by setting the LIBS environment variable:

~~~
LIBS='-L/X/netcdf-4.0/lib -lnetcdf -L/X/hdf5-1.8.9/lib -lhdf5_hl -lhdf5 -lz -lm -L/X/szip-2.1/lib -lsz'
~~~

For shared builds, only -lnetcdf is needed. All other libraries will
be found automatically.

The ``nc-config --all'' command can be used to learn what options are
needed for the local netCDF installation.

For example, this works for linking an application named myapp.c with
netCDF-4 libraries:

~~~
cc -o myapp myapp.c `nc-config --cflags --libs`
~~~

configure options {#configure_options}
-----------------------------

These options are used for `autotools`-based builds.  For `cmake` options, see

Note: --disable prefix indicates that the option is normally enabled.
<table>
<tr><th>Option<th>Description<th>Dependencies
<tr><td>--disable-doxygen<td>Disable generation of documentation.<td>doxygen
<tr><td>--disable-fsync<td>disable fsync support<td>kernel fsync support
<tr><td>--enable-valgrind-tests <td>build with valgrind-tests; static builds only<td>valgrind
<tr><td>--enable-netcdf-4<td>build with netcdf-4<td>HDF5 and zlib
<tr><td>--enable-netcdf4<td>synonym for enable-netcdf-4
<tr><td>--enable-hdf4<td>build netcdf-4 with HDF4 read capability<td>HDF4, HDF5 and zlib
<tr><td>--enable-hdf4-file-tests<td>test ability to read HDF4 files<td>selected HDF4 files from Unidata ftp site
<tr><td>--enable-pnetcdf<td>build netcdf-4 with parallel I/O for classic and
                          64-bit offset files using parallel-netcdf
<tr><td>--enable-extra-example-tests<td>Run extra example tests<td>--enable-netcdf-4,GNU sed
<tr><td>--enable-parallel-tests <td>run extra parallel IO tests<td>--enable-netcdf-4, parallel IO support
<tr><td>--enable-logging<td>enable logging capability<td>--enable-netcdf-4
<tr><td>--disable-dap<td>build without DAP client support.<td>libcurl
<tr><td>--disable-dap-remote-tests<td>disable dap remote tests<td>--enable-dap
<tr><td>--enable-dap-long-tests<td>enable dap long tests<td>
<tr><td>--enable-extra-tests<td>run some extra tests that may not pass because of known issues<td>
<tr><td>--enable-ffio<td>use ffio instead of posixio (ex. on the Cray)<td>
<tr><td>--disable-examples<td>don't build the netCDF examples during make check
                          (examples are treated as extra tests by netCDF)<td>
<tr><td>--disable-v2<td>turn off the netCDF version 2 API<td>
<tr><td>--disable-utilities<td>don't build netCDF utilities ncgen, ncdump, and nccopy<td>
<tr><td>--disable-testsets<td>don't build or run netCDF tests<td>
<tr><td>--enable-large-file-tests <td>Run tests which create very large data
		          files<td>~13 GB disk space required, but recovered when
                          tests are complete). See option --with-temp-large to
                          specify temporary directory<td>
<tr><td>--enable-benchmarks<td>Run benchmarks. This is an experimental feature.
			  The benchmarks are a
                          bunch of extra tests, which are timed. We use these
                          tests to check netCDF performance.
    <td>sample data files from the Unidata ftp site
<tr><td>--disable-extreme-numbers
<td>don't use extreme numbers during testing, such as MAX_INT - 1<td>
<tr><td>--enable-dll<td>build a win32 DLL<td>mingw compiler
<tr><td>--disable-shared<td>build shared libraries<td>
<tr><td>--disable-static<td>build static libraries<td>
<tr><td>--disable-largefile<td>omit support for large files<td>
<tr><td>--enable-mmap<td>Use mmap to implement NC_DISKLESS<td>
</table>

Build Instructions for netCDF-C using CMake {#netCDF-CMake}
===========================================

## Overview {#cmake_overview}

Starting with netCDF-C 4.3.0, we are happy to announce the inclusion of CMake support.  CMake will allow for building netCDF on a wider range of platforms, include Microsoft Windows with Visual Studio.  CMake support also provides robust unit and regression testing tools.  We will also maintain the standard autotools-based build system in parallel.

In addition to providing new build options for netCDF-C, we will also provide pre-built binary downloads for the shared versions of netCDF for use with Visual Studio.


##  Requirements {#cmake_requirements}
The following packages are required to build netCDF-C using CMake.

* netCDF-C Source Code
* CMake version 2.8.12 or greater.
* Optional Requirements:
	* HDF5 Libraries for netCDF4/HDF5 support.
	* libcurl for DAP support.

<center>
<img src="deptree.jpg" height="250px" />
</center>

## The CMake Build Process {#cmake_build}

There are four steps in the Build Process when using CMake

1. Configuration: Before compiling, the software is configured based on the desired options.
2. Building: Once configuration is complete, the libraries are compiled.
3. Testing: Post-build, it is possible to run tests to ensure the functionality of the netCDF-C libraries.
4. Installation: If all tests pass, the libraries can be installed in the location specified during configuration.

For users who prefer pre-built binaries, installation packages are available at \ref winbin

### Configuration {#cmake_configuration}

The output of the configuration step is a project file based on the appropriate configurator specified.  Common configurators include:

* Unix Makefiles
* Visual Studio
* CodeBlocks
* ... and others

### Common CMake Options {#cmake_common_options}

| **Option** | **Autotools** | **CMake** |
| :------- | :---- | :----- |
Specify Install Location | --prefix=PREFIX | -D"CMAKE\_INSTALL\_PREFIX=PREFIX"
Enable/Disable netCDF-4 | --enable-netcdf-4<br>--disable-netcdf-4 | -D"ENABLE\_NETCDF\_4=ON" <br> -D"ENABLE\_NETCDF\_4=OFF"
Enable/Disable DAP | --enable-dap <br> --disable-dap | -D"ENABLE\_DAP=ON" <br> -D"ENABLE\_DAP=OFF"
Enable/Disable Utilities | --enable-utilities <br> --disable-utilities | -D"BUILD\_UTILITIES=ON" <br> -D"BUILD\_UTILITIES=OFF"
Specify shared/Static Libraries | --enable-shared <br> --enable-static | -D"BUILD\_SHARED\_LIBS=ON" <br> -D"BUILD\_SHARED\_LIBS=OFF"
Enable/Disable Tests | --enable-testsets <br> --disable-testsets | -D"ENABLE\_TESTS=ON" <br> -D"ENABLE\_TESTS=OFF"
Specify a custom library location | Use *CFLAGS* and *LDFLAGS* | -D"CMAKE\_PREFIX\_PATH=/usr/custom_libs/"

A full list of *basic* options can be found by invoking `cmake [Source Directory] -L`. To enable a list of *basic* and *advanced* options, one would invoke `cmake [Source Directory] -LA`.

### Configuring your build from the command line. {#cmake_command_line}

The easiest configuration case would be one in which all of the dependent libraries are installed on the system path (in either Unix/Linux or Windows) and all the default options are desired. From the build directory (often, but not required to be located within the source directory):

> $ cmake [Source Directory]

If you have libraries installed in a custom directory, you may need to specify the **CMAKE\_PREFIX_PATH** variable to tell cmake where the libraries are installed. For example:

> $ cmake [Source Directory] -DCMAKE\_PREFIX\_PATH=/usr/custom_libraries/

## Building {#cmake_building}

The compiler can be executed directly with 'make' or the appropriate command for the configurator which was used.

> $ make

Building can also be executed indirectly via cmake:

> $ cmake --build [Build Directory]

## Testing {#cmake_testing}

Testing can be executed several different ways:

> $ make test

or

> $ ctest

or

> $ cmake --build [Build Directory] --target test

### Installation {#cmake_installation}

Once netCDF has been built and tested, it may be installed using the following commands:

> $ make install

or

> $ cmake --build [Build Directory] --target install

## See Also {#cmake_see_also}

For further information regarding netCDF and CMake, see \ref cmake_faq

**Comment by jbeezley** [comment]
I think I'm having the same issue, but I'm not using cmake.  See the [test log](https://gist.github.com/jbeezley/750d10549dfe162c0cdf).  Run manually, the tests say they pass.

Also after installing version 4.3.3, `ncdump -h <url>` returns ` NetCDF: I/O failure` for every DAP url I have tried.  The same url's work with version 4.3.2.

**Comment by durack1** [comment]
@DennisHeimbigner would I be right to assume that this issue is a duplicate of #117?

**Comment by WardF** [comment]
I believe that this is a duplicate of #117 as previous mentioned; closing.

**Comment by durack1** [comment]
@WardF, yep you're right to tag this as a dupe.. Hopefully the root cause of the problem can be isolated and fixed as part of #117
